File: app/src/main/java/co/tpcreative/supersafe/ui/albumdetail/AlbumDetailAdapter.java
package co.tpcreative.supersafe.ui.albumdetail;
import android.content.Context;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;
import com.bumptech.glide.Glide;
import com.bumptech.glide.Priority;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.ImageViewTarget;
import com.snatik.storage.Storage;
import butterknife.BindView;
import butterknife.OnClick;
import butterknife.OnLongClick;
import co.tpcreative.supersafe.R;
import co.tpcreative.supersafe.common.adapter.BaseAdapter;
import co.tpcreative.supersafe.common.adapter.BaseHolder;
import co.tpcreative.supersafe.common.entities.ItemEntity;
import co.tpcreative.supersafe.common.services.SuperSafeApplication;
import co.tpcreative.supersafe.common.util.Utils;
import co.tpcreative.supersafe.model.EnumFormatType;
import co.tpcreative.supersafe.model.EnumStatusProgress;
import co.tpcreative.supersafe.model.ItemModel;
import co.tpcreative.supersafe.model.ThemeApp;

public class AlbumDetailAdapter extends BaseAdapter<ItemModel, BaseHolder> {
    RequestOptions options = new RequestOptions()
            .centerCrop()
            .override(200 ,200)
            .placeholder(R.color.material_gray_100)
            .error(R.color.red_200)
            .priority(Priority.HIGH);
    private Context context;
    private ItemSelectedListener itemSelectedListener;
    private Storage storage;
    private String TAG = AlbumDetailAdapter.class.getSimpleName();
    final ThemeApp themeApp = ThemeApp.getInstance().getThemeInfo();
    Drawable note1 = SuperSafeApplication.getInstance().getResources().getDrawable( themeApp.getAccentColor());

    public AlbumDetailAdapter(LayoutInflater inflater, Context context, ItemSelectedListener itemSelectedListener) {
        super(inflater);
        this.context = context;
        storage = new Storage(context);
        this.itemSelectedListener = itemSelectedListener;
        storage.setEncryptConfiguration(SuperSafeApplication.getInstance().getConfigurationFile());
    }

    @Override
    public int getItemCount() {
        return dataSource.size();
    }

    @Override
    public BaseHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ItemHolder(inflater.inflate(R.layout.album_detail_item, parent, false));
    }


    public interface ItemSelectedListener {
        void onClickItem(int position);
        void onLongClickItem(int position);
    }

    public class ItemHolder extends BaseHolder<ItemModel> {
        @BindView(R.id.imgAlbum)
        ImageView imgAlbum;
        @BindView(R.id.imgVideoCam)
        ImageView imgVideoCam;
        @BindView(R.id.tvTitle)
        TextView tvTitle;
        @BindView(R.id.progressingBar)
        ProgressBar progressingBar;
        @BindView(R.id.imgCheck)
        ImageView imgCheck;
        @BindView(R.id.view_alpha)
        View view_alpha;
        @BindView(R.id.imgSelect)
        ImageView imgSelect;
        int mPosition;
        @BindView(R.id.rlHome)
        RelativeLayout rlHome;
        public ItemHolder(View itemView) {
            super(itemView);
        }

        @Override
        public void bind(ItemModel data, int position) {
            super.bind(data, position);
            mPosition = position;
            Utils.Log(TAG,"Position "+ position);
            if (data.isChecked) {
                view_alpha.setAlpha(0.5f);
                imgSelect.setVisibility(View.VISIBLE);
            } else {
                view_alpha.setAlpha(0.0f);
                imgSelect.setVisibility(View.INVISIBLE);
            }
            try {
                String path = data.thumbnailPath;
                EnumFormatType formatTypeFile = EnumFormatType.values()[data.formatType];
                switch (formatTypeFile) {
                    case AUDIO: {
                        imgVideoCam.setVisibility(View.VISIBLE);
                        imgVideoCam.setImageDrawable(context.getResources().getDrawable(R.drawable.baseline_music_note_white_48));
                        tvTitle.setVisibility(View.VISIBLE);
                        tvTitle.setText(data.title);
                        Glide.with(context)
                                .load(note1)
                                .apply(options).into(imgAlbum);
                        break;
                    }
                    case VIDEO: {
                        imgVideoCam.setVisibility(View.VISIBLE);
                        imgVideoCam.setImageDrawable(context.getResources().getDrawable(R.drawable.baseline_videocam_white_36));
                        tvTitle.setVisibility(View.INVISIBLE);
                        if (storage.isFileExist(path)){
                            imgAlbum.setRotation(data.degrees);
                            Glide.with(context)
                                    .load(storage.readFile(path))
                                    .apply(options).into(imgAlbum);
                        }
                        break;
                    }
                    case IMAGE: {
                        tvTitle.setVisibility(View.INVISIBLE);
                        imgVideoCam.setVisibility(View.INVISIBLE);
                        if (storage.isFileExist(path)){
                            imgAlbum.setRotation(data.degrees);
                            Glide.with(context)
                                    .load(storage.readFile(path))
                                    .apply(options).into(imgAlbum);
                        }
                        break;
                    }
                    case FILES:{
                        imgVideoCam.setVisibility(View.VISIBLE);
                        imgVideoCam.setImageDrawable(context.getResources().getDrawable(R.drawable.baseline_insert_drive_file_white_48));
                        tvTitle.setVisibility(View.VISIBLE);
                        tvTitle.setText(data.title);
                        Glide.with(context)
                                .load(note1)
                                .apply(options).into(imgAlbum);
                        break;
                    }
                }

                progressingBar.getIndeterminateDrawable().setColorFilter(context.getResources().getColor(themeApp.getAccentColor()),
                        PorterDuff.Mode.SRC_IN);
                EnumStatusProgress progress = EnumStatusProgress.values()[data.statusProgress];
                switch (progress){
                    case PROGRESSING:{
                        imgCheck.setVisibility(View.INVISIBLE);
                        progressingBar.setVisibility(View.VISIBLE);
                        break;
                    }
                    case DONE:{
                        if (data.isSaver){
                            imgCheck.setImageDrawable(context.getResources().getDrawable(R.drawable.baseline_attach_money_white_48));
                            imgCheck.setVisibility(View.VISIBLE);
                        }
                        else{
                            imgCheck.setVisibility(View.VISIBLE);
                        }
                        progressingBar.setVisibility(View.INVISIBLE);
                        break;
                    }
                    default:{
                        imgCheck.setVisibility(View.INVISIBLE);
                        progressingBar.setVisibility(View.INVISIBLE);
                        break;
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        @OnClick(R.id.rlHome)
        public void onClicked(View view) {
            if (itemSelectedListener != null) {
                itemSelectedListener.onClickItem(mPosition);
            }
        }

        @OnLongClick(R.id.rlHome)
        public boolean onLongClickedItem(View view){
            if (itemSelectedListener!=null){
                itemSelectedListener.onLongClickItem(mPosition);
            }
            return true;
        }
    }
}

File: app/src/main/java/co/tpcreative/supersafe/ui/albumdetail/AlbumDetailAdapter.kt
package co.tpcreative.supersafe.ui.albumdetail
import android.annotation.SuppressLint
import android.content.Context
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.AppCompatTextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.GridLayoutManager
import co.tpcreative.supersafe.R
import co.tpcreative.supersafe.common.adapter.BaseAdapter
import co.tpcreative.supersafe.common.adapter.BaseHolder
import co.tpcreative.supersafe.common.extension.isFileExist
import co.tpcreative.supersafe.common.extension.readFile
import co.tpcreative.supersafe.common.services.SuperSafeApplication
import co.tpcreative.supersafe.common.util.Utils
import co.tpcreative.supersafe.model.*
import com.bumptech.glide.Glide
import com.bumptech.glide.Priority
import com.bumptech.glide.request.RequestOptions
import kotlinx.android.synthetic.main.album_detail_item.view.*
import kotlinx.android.synthetic.main.album_detail_item.view.imgAlbum
import kotlinx.android.synthetic.main.album_detail_item.view.rlHome
import kotlinx.android.synthetic.main.album_detail_item.view.tvTitle
import kotlinx.android.synthetic.main.album_detail_item.view.view_alpha
import kotlinx.android.synthetic.main.custom_item_verical.view.*
import tpcreative.co.qrscanner.common.extension.setColorFilter

class AlbumDetailAdapter(private val mLayoutManager: GridLayoutManager? = null, inflater: LayoutInflater, private val context: Context?, private val itemSelectedListener: ItemSelectedListener?) : BaseAdapter<ItemModel, BaseHolder<ItemModel>>(inflater) {
    var options: RequestOptions? = RequestOptions()
            .centerCrop()
            .override(200, 200)
            .placeholder(R.color.material_gray_100)
            .error(R.color.red_200)
            .priority(Priority.HIGH)
    private val TAG = AlbumDetailAdapter::class.java.simpleName
    val themeApp: ThemeApp? = ThemeApp.getInstance()?.getThemeInfo()
    var note1: Drawable? = ContextCompat.getDrawable(SuperSafeApplication.getInstance(), themeApp!!.getAccentColor())

    companion object {
        const val SPAN_COUNT_ONE = 1
        const val SPAN_COUNT_THREE = 3
    }

    private val VIEW_TYPE_SMALL = 1
    private val VIEW_TYPE_BIG = 2

    override fun getItemCount(): Int {
        return mDataSource.size
    }


    override fun getItemViewType(position: Int): Int {
        val spanCount = mLayoutManager?.spanCount
        return if (spanCount == SPAN_COUNT_THREE) {
           VIEW_TYPE_BIG
        } else {
           VIEW_TYPE_SMALL
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseHolder<ItemModel> {
        if (viewType==VIEW_TYPE_SMALL){
            return ItemHolderVertical(inflater!!.inflate(R.layout.custom_item_verical, parent, false))
        }
        return ItemHolder(inflater!!.inflate(R.layout.album_detail_item, parent, false))
    }

    interface ItemSelectedListener {
        fun onClickItem(position: Int)
        fun onLongClickItem(position: Int)
    }

    inner class ItemHolder(view: View) : BaseHolder<ItemModel>(view) {

        val imgAlbum = itemView.imgAlbum


        val imgVideoCam = itemView.imgVideoCam

        val tvTitle = itemView.tvTitle

        val progressingBar = itemView.progressingBar

        val imgCheck = itemView.imgCheck

        val view_alpha = itemView.view_alpha

        val imgSelect = itemView.imgSelect
        var mPosition = 0

        val rlHome = itemView.rlHome
        override fun bind(data: ItemModel, position: Int) {
            super.bind(data, position)
            mPosition = position
            if (data.isChecked) {
                view_alpha?.alpha = 0.5f
                imgSelect?.visibility = View.VISIBLE
                Utils.Log(TAG, "VISIBLE")
            } else {
                view_alpha?.alpha = 0.0f
                imgSelect?.visibility = View.INVISIBLE
            }
            try {
                val path: String? = data.getThumbnail()
                when (EnumFormatType.values()[data.formatType]) {
                    EnumFormatType.AUDIO -> {
                        imgVideoCam?.visibility = View.VISIBLE
                        imgVideoCam?.setImageDrawable(ContextCompat.getDrawable(context!!, R.drawable.baseline_music_note_white_48))
                        tvTitle?.visibility = View.VISIBLE
                        tvTitle?.text = data.title
                        Glide.with(context!!)
                                .load(note1)
                                .apply(options!!).into(imgAlbum!!)
                    }
                    EnumFormatType.VIDEO -> {
                        imgVideoCam?.visibility = View.VISIBLE
                        imgVideoCam?.setImageDrawable(ContextCompat.getDrawable(context!!, R.drawable.baseline_videocam_white_36))
                        tvTitle?.visibility = View.INVISIBLE
                        if (path?.isFileExist()!!) {
                            imgAlbum?.rotation = data.degrees.toFloat()
                            Glide.with(context!!)
                                    .load(path.readFile())
                                    .apply(options!!).into(imgAlbum!!)
                        }
                    }
                    EnumFormatType.IMAGE -> {
                        tvTitle?.visibility = View.INVISIBLE
                        imgVideoCam?.visibility = View.INVISIBLE
                        if (path!!.isFileExist()) {
                            imgAlbum?.rotation = data.degrees.toFloat()
                            Glide.with(context!!)
                                    .load(path.readFile())
                                    .apply(options!!).into(imgAlbum!!)
                        }
                    }
                    EnumFormatType.FILES -> {
                        imgVideoCam?.visibility = View.VISIBLE
                        imgVideoCam?.setImageDrawable(ContextCompat.getDrawable(context!!, R.drawable.baseline_insert_drive_file_white_48))
                        tvTitle?.visibility = View.VISIBLE
                        tvTitle?.text = data.title
                        Glide.with(context!!)
                                .load(note1)
                                .apply(options!!).into(imgAlbum!!)
                    }
                }
                progressingBar?.indeterminateDrawable?.setColorFilter(ContextCompat.getColor(context!!, themeApp!!.getAccentColor()), EnumMode.SRC_ATOP)
                when (EnumStatusProgress.values()[data.statusProgress]) {
                    EnumStatusProgress.PROGRESSING -> {
                        imgCheck?.visibility = View.INVISIBLE
                        progressingBar?.visibility = View.VISIBLE
                    }
                    EnumStatusProgress.DONE -> {
                        if (data.isSaver) {
                            imgCheck?.setImageDrawable(ContextCompat.getDrawable(context!!, R.drawable.baseline_attach_money_white_48))
                            imgCheck?.visibility = View.VISIBLE
                        } else {
                            imgCheck?.visibility = View.VISIBLE
                        }
                        progressingBar?.visibility = View.INVISIBLE
                    }
                    else -> {
                        imgCheck?.visibility = View.INVISIBLE
                        progressingBar?.visibility = View.INVISIBLE
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
            rlHome.setOnClickListener {
                itemSelectedListener?.onClickItem(mPosition)
            }

            rlHome.setOnLongClickListener {
                itemSelectedListener?.onLongClickItem(mPosition)
                true
            }
        }
    }

    inner class ItemHolderVertical(itemView: View) : BaseHolder<ItemModel>(itemView) {
        val imgAlbum: AppCompatImageView = itemView.imgAlbumV
        val tvTitle: AppCompatTextView = itemView.tvTitleV
        val tvSizeCreatedDate: AppCompatTextView = itemView.tvSizeCreatedDateV
        val imgIcon : AppCompatImageView = itemView.imgIcon
        var alpha = itemView.view_alphaV
        var mPosition = 0
        @SuppressLint("SetTextI18n")
        override fun bind(data: ItemModel, position: Int) {
            super.bind(data, position)
            mPosition = position
            Utils.Log(TAG, "Position $position")
            if (data.isChecked) {
                alpha?.alpha = 0.5f
            } else {
                alpha?.alpha = 0.0f
            }
            try {
                val path: String = data.getThumbnail()
                val formatTypeFile = EnumFormatType.values()[data.formatType]
                tvTitle.text = data.titleView
                tvSizeCreatedDate.text = data.createdDateTimeView
                when (formatTypeFile) {
                    EnumFormatType.AUDIO -> {
                        Glide.with(context!!)
                                .load(note1)
                                .apply(options!!).into(imgAlbum)
                        imgIcon.background = ContextCompat.getDrawable(context,R.drawable.baseline_music_note_white_48)
                        imgIcon.visibility = View.VISIBLE
                    }
                    EnumFormatType.VIDEO -> {
                        if (path.isFileExist()) {
                            imgAlbum.rotation = data.degrees.toFloat()
                            Glide.with(context!!)
                                    .load(path.readFile())
                                    .apply(options!!).into(imgAlbum)
                        }
                        imgIcon.visibility = View.INVISIBLE
                    }
                    EnumFormatType.IMAGE -> {
                        if (path.isFileExist()) {
                            imgAlbum.rotation = data.degrees.toFloat()
                            Glide.with(context!!)
                                    .load(path.readFile())
                                    .apply(options!!).into(imgAlbum)
                        }
                        imgIcon.visibility = View.INVISIBLE
                    }
                    EnumFormatType.FILES -> {
                        Glide.with(context!!)
                                .load(note1)
                                .apply(options!!).into(imgAlbum)
                        imgIcon.background = ContextCompat.getDrawable(context,R.drawable.baseline_insert_drive_file_white_48)
                        imgIcon.visibility = View.VISIBLE
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }

            itemView.rlHomeV.setOnClickListener {
                itemSelectedListener?.onClickItem(mPosition)
            }
            itemView.rlHomeV.setOnLongClickListener {
                itemSelectedListener?.onLongClickItem(mPosition)
                true
            }
        }
    }

}